google/cloud/pubsub/internal/extend_leases_with_retry.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/pubsub/internal/extend_leases_with_retry.h"
+#include "google/cloud/pubsub/internal/exactly_once_policies.h"
+#include "google/cloud/completion_queue.h"
+#include "google/cloud/log.h"
+#include <algorithm>
+#include <memory>
+
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+// This value is hard-coded in the spec.
+auto constexpr kMaxRetryAttempts = 3;
+
+/// Updates a ModifyAckDeadlineRequest request to drop ack_ids with permanent
+/// failures, and successful ack_ids.
+google::pubsub::v1::ModifyAckDeadlineRequest UpdateRequest(
+    google::pubsub::v1::ModifyAckDeadlineRequest request,
+    Status const& status) {
+  if (ExactlyOnceRetryable(status.code())) return request;
+
+  auto permanent_error = [](std::string const& ack_id, Status const& status) {
+    auto const& m = status.error_info().metadata();
+    auto f = m.find(ack_id);
+    auto const permanent =
+        f == m.end() || f->second.rfind("TRANSIENT_FAILURE_", 0) != 0;
+    if (permanent) {
+      GCP_LOG(WARNING)
```

Comment:
Not a big deal, but there is nothing to control whether this operation logs or not. You could introduce Options and check for some new `TracingComponentsOption`. but eh.

Reply:
Nevermind. It looks like we have a lot of things that log outside of our logging classes

Reply:
Ack, but I think you now agree this is fine.